### PR TITLE
Remove warnings when used on latest rust nightly

### PR DESCRIPTION
**Pull request by azdlowry**
Latest nightly has improved the lint for unused muts, meaning a lot of extra warnings when using chan.

This PR removes the unused muts and turns on warnings as error to prevent warnings coming back in the future.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -394,6 +394,7 @@ authors has done substantial research on the integration of CSP and programming
 languages.
 */
 
+#![deny(warnings)]
```
Comment (BurntSushi) [inline src/lib.rs]: I admire your gusto, but we shouldn't fail the build because there are warnings. If new warnings come out again, then `chan` will stop compiling and that's worse than spewing warnings but still compiling.

File: Cargo.toml [hunk 2]
```diff
@@ -11,7 +11,7 @@ keywords = ["channel", "asynchronous", "mpmc", "golang", "select"]
 license = "Unlicense/MIT"
 
 [dependencies]
-rand = "0.3"
+rand = "0.4"
```
Comment (BurntSushi) [inline Cargo.toml]: You'll need to remove this. While I don't have the Travis CI configured right, this is going to bump the minimum Rust version required by `chan`, which will in turn require me to do a semver bump of `chan`. I'm not sure that I'm ready to do that.

File: src/lib.rs (hunk 1 above)
Comment (BurntSushi) [reply src/lib.rs]: As a compromise, you could do `#![cfg_attr(test, deny(warnings))]`, which will fail the build only in tests.

File: examples/web.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
+#![deny(warnings)]
```
Comment (BurntSushi) [inline examples/web.rs]: Please remove these from all of the examples. I'd rather folks be able to compile examples even if there are warnings.

File: README.md [hunk 4]
```diff
@@ -131,3 +131,4 @@ for using channels in `chan` to receive OS signals. (Unix only at the moment.)
 [cmail](https://github.com/BurntSushi/rust-cmail) is a program to periodically
 email the output of long running commands. I ported this from an
 [earlier version that I wrote in Go](https://github.com/BurntSushi/cmail).
+
```
Comment (BurntSushi) [inline README.md]: Please remove this extra new line.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Thanks! I have a few nits. After those are resolved, this should be good to merge.

File: Cargo.toml (hunk 2 above)
Comment (azdlowry) [reply Cargo.toml]: There will have been a very good reason for me doing this, as it's not something I would do likely, but unfortunately I've forgotten what it is. I'll see if I can dig up the reason.

**Comment by azdlowry** [comment]
OK, I understand your points. Some background might help us to come a better solution. Currently we've not been able to use chan crate on crates.io for about 5 months, since moving to a more recent rust version. The reason is this:

error: variable does not need to be mutable
   --> src/netopt/src/non_blocking_conn.rs:135:9
    |
135 | /         chan_select! {
136 | |             default => { },
137 | |             addr_rx.recv() -> next => { result = next }
138 | |         }
    | |_________^ help: remove this `mut`
    |

We have '#![deny(warnings)]' on all our code and everywhere we use the 'chan_select!' macro we have a build failure. So some warnings in chan cause build failures for us.

I think it would be good if we could catch these issues further up stream, so a minor version bump of chan couldn't cause build failures in downstream code.

Maybe we could be very specific about where we put 'deny(warnings)'. What do you think?

**Comment by BurntSushi** [comment]
@azdlowry Yeah I mean, I do not want to be sprinkling `deny(warnings)` around the source code for the same reason I don't want it at the top level. New warnings can and will be introduced, and I do not want `chan` to stop compiling for _everyone_ because of that. Could you say why enabling `deny(warnings)` only when running tests isn't sufficient for you? If new warnings are introduced, then new builds will fail, so I'm not sure what an unconditional `deny(warnings)` gets us (other than causing unnecessary compilation failures).

It's unfortunate that these warnings bleed through the `chan_select!` macro, and that's definitely a corner case, but I think I stand by my comments above.

**Comment by azdlowry** [comment]
As 'chan_select!' is used in the tests it would fail in travis if a change caused a warning, so that's a good approach.  I'll look at changing it to that.

**Comment by azdlowry** [comment]
'#![cfg_attr(test, deny(warnings))]' doesn't work, and I don't have time to investigate why at the moment. Thinking back that might be why I put them in the examples.

**Comment by BurntSushi** [comment]
Thanks! I'll merge this for now and take a look at adding `#![cfg_attr(test, deny(warnings))]`. I've used it in other crates so it should definitely work!
